File: scripts/debug_strides.py
import torch
from safetensors.torch import save_file
import numpy as np

def debug_strides():
    x = torch.zeros(1, 2, 32, 32, 32)
    x[0, 0, 16, 16, 16] = 1.0
    x[0, 1, 16, 16, 16] = 1.0
    
    print("Shape:", x.shape)
    print("Stride:", x.stride())
    print("Is contiguous:", x.is_contiguous())
    
    # Flatten
    flat = x.view(-1)
    indices = torch.nonzero(flat).flatten()
    print("Linear indices of 1.0:", indices.tolist())
    
    # Save as numpy too for cross-check
    np.save("verification_data/x_debug.npy", x.numpy())
    save_file({"input": x}, "verification_data/strides.safetensors")

if __name__ == "__main__":
    debug_strides()

File: scripts/debug_strides.jl
using SafeTensors

function debug_strides_jl()
    tensors = SafeTensors.load_safetensors("verification_data/strides.safetensors")
    x = tensors["input"]
    println("Size: ", size(x))
    
    # Linear indices
    inds = findall(v -> abs(v) > 0.5, x)
    linear_inds = [LinearIndices(x)[i] for i in inds]
    println("Linear Indices (1-based): ", linear_inds)
    println("Linear Indices (0-based): ", linear_inds .- 1)
    
    # Check expected 16912, 49680
    expected = [16912, 49680]
    println("Match expected (0-based)? ", sort(linear_inds .- 1) == sort(expected))
end

debug_strides_jl()
